### ISSUE TITLE
Tracking Issue for NEON intrinsics on 32-bit ARM

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/rust#111800]
Feature gate: `#![feature(stdarch_arm_neon_intrinsics)]`

This is a tracking issue for the NEON intrinsics in `core::arch::arm`. This was split off from #90972 which tracked the AArch64 NEON intrinsics which have now been stabilized.

### Public API

Everything in `core::arch::arm` that starts with the letter `v`.

### Steps / History

- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by RalfJung** [comment]
> In #126555 all SIMD types have been changed to require `neon` to be enabled (even if the underlying register does not) as LLVM fails to compile using them as ASM inputs/outputs otherwise.

So what happens when I use the type and don't have NEON enabled?
I can't see anything in that PR that would then emit an error or so.

**Comment by RalfJung** [comment]
Note that this feature is currently unsound due to https://github.com/rust-lang/rust/issues/129880.

**Comment by RalfJung** [comment]
Ah, makes sense. :)

Once https://github.com/rust-lang/rust/pull/127731 lands, we should probably also require NEON when passing vectors to/from non-Rust-ABI functions.

**Comment by Nugine** [comment]
Another unsoundness due to codegen: https://github.com/rust-lang/stdarch/issues/1217

**Comment by Jamesbarford** [comment]
I noticed that the concerns around NEON requirements for inline assembly and non-Rust ABI functions have been addressed, and #127731 has already landed.

Given that the API of these intrinsics will not change further; I've prepared a PR (which I haven't yet opened) to move forward with stabilisation. The PR also includes improvements to the documentation of these intrinsics. Would it be possible to move forward with this?

The soundness issue appears to be endemic to _any_ platform, possibly even any compiler using LLVM as a backend as the same issue is observed in C/C++ when using LLVM. Given this, it seems to be a separate problem from stabilizing the API of these intrinsics.

**Comment by Amanieu** [comment]
The first thing that needs to be done before any stabilization here is cleaning up and stabilizing the target features for AArch32. This will be needed for any of these intrinsics to actually be called.

**Comment by RalfJung** [comment]
> The soundness issue appears to be endemic to any platform

No, this is an ARM-32 issue. All other platforms implement floating-points properly, without mandatory subnormal flushing.

**Comment by RalfJung** [comment]
> changing the rounding mode of floats appears to cause a similar issue regardless of Arm-32

Sure, if you call the deprecated, unsafe, and documented-to-be-unsound operations that change the rounding mode then you get in trouble. That's multiple light years away from "I used a normal, maybe even safe NEON intrinsic and got unsoundness".

> I thought the stabilization process was more about the API design rather than its implementation?

Clearly we wouldn't stabilize something whose implementation is incomplete or buggy, no matter how sure we are about the API.
